docs/proposals/templates.md

```diff
@@ -0,0 +1,550 @@
+* Support for non-string parameters.  K8S objects define fields of type string, int, bool, and byte[](base64).  The parameter
+  scheme should support all of those types for substitution.
+  * complex types (struct/map/list) can be defined in terms of the available primitives, so it's preferred to avoid the complexity
+    of allowing for full complex-type substitution.
+* Parameter metadata.  Parameters should include at a minimum, information describing the purpose of the parameter, whether it is 
+  required/optional, and a default/suggested value.  Type information could also be required to resolve the non-string substitution
+  requirement.
+* Template metadata.  Templates should be able to include metadata describing their purpose or links to further documentation and 
+  versioning information.
+
+
+## Proposed Implementation
+
+### Overview
+
+We began by looking at the List object which allows a user to easily group a set of objects together for easy creation via a 
+single CLI invocation.  It also provides a portable format which requires only a single file to represent an application.
+
+From that starting point, we propose a Template API object which can encapsulate the definition of all components of an 
+application to be created.  The application definition is encapsulated in the form of an array of API objects (identical to 
+List), plus a parameterization section.  Components reference the parameter by name and the value of the parameter is 
+substituted during a processing step, prior to submitting each component to the appropriate API endpoint for creation.
+
+The primary capability provided is that parameter values can easily be shared between components, such as a database password 
+that is provided by the user once, but then attached as an environment variable to both a database pod and a web frontend pod.
+
+In addition, the template can be repeatedly instantiated for a consistent application deployment experience in different 
+namespaces or Kubernetes clusters.
+
+Lastly, we propose the Template API object include a “Labels” section in which the template author can define a set of labels 
+to be applied to all objects created from the template.  This will give the template deployer an easy way to manage all the 
+components created from a given template.  These labels will also be applied to selectors defined by Objects within the template,
+allowing a combination of templates and labels to be used to scope resources within a namespace.  That is, a given template
+can be instantiated multiple times within the same namespace, as long as a different label value is used each for each
+instantiation.  The resulting objects will be independent from a replica/load-balancing perspective.
+
+Generation of parameter values for fields such as Secrets will be delegated to an [admission controller/initializer/finalizer](https://github.com/kubernetes/kubernetes/issues/3585) rather than being solved by the template processor.
```

Comment:
I've also previously proposed that our controllers could (optionally) generate unique label values: #12298.